precompiles/dispatch-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use alloc::format;
+use core::marker::PhantomData;
+use fp_evm::{ExitError, PrecompileFailure, PrecompileHandle};
+use frame_support::pallet_prelude::IsType;
+use frame_support::{codec::DecodeLimit as _, dispatch::Pays, traits::Get};
+use frame_support::{
+    dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+    traits::ConstU32,
+};
+use frame_system::Config;
+use pallet_evm::GasWeightMapping;
+use pallet_evm_precompile_dispatch::DispatchValidateT;
+use precompile_utils::prelude::{revert, BoundedBytes, UnboundedBytes};
+use precompile_utils::EvmResult;
+use sp_core::{crypto::AccountId32, H160, H256};
+use sp_io::hashing::keccak_256;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const LOG_TARGET: &str = "precompile::dispatch-lockdrop";
+
+// ECDSA PublicKey
+type ECDSAPublic = ConstU32<64>;
+
+// `DecodeLimit` specifies the max depth a call can use when decoding, as unbounded depth
+// can be used to overflow the stack.
+// Default value is 8, which is the same as in XCM call decoding.
+pub struct DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit = ConstU32<8>>(
+    PhantomData<(Runtime, DispatchValidator, DecodeLimit)>,
+);
+
+#[precompile_utils::precompile]
+impl<Runtime, DispatchValidator, DecodeLimit>
+    DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit>
+where
+    Runtime: pallet_evm::Config,
+    <Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+    Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+    <Runtime as Config>::AccountId: IsType<AccountId32>,
+    <Runtime as Config>::AccountId: From<[u8; 32]>,
+    DispatchValidator:
+        DispatchValidateT<<Runtime as Config>::AccountId, <Runtime as Config>::RuntimeCall>,
+    DecodeLimit: Get<u32>,
+{
+    #[precompile::public("dispatch_lockdrop_call(bytes,bytes)")]
+    fn dispatch_lockdrop_call(
+        handle: &mut impl PrecompileHandle,
+        call: UnboundedBytes,
+        pubkey: BoundedBytes<ECDSAPublic>,
+    ) -> EvmResult<bool> {
+        log::trace!(
+            target: LOG_TARGET,
+            "raw arguments: call: {:?}, pubkey: {:?}",
+            call,
+            pubkey
+        );
+
+        let caller: H160 = handle.context().caller.into();
+        let input: Vec<u8> = call.into();
+
+        // 1. Decode the call
+        let call = Runtime::RuntimeCall::decode_with_depth_limit(DecodeLimit::get(), &mut &*input)
+            .map_err(|_| revert("could not decode call"))?;
+
+        // 2. Check if dispatching the call will not exceed the gas limit
+        let mut gas_limit = handle.remaining_gas();
+        // If caller specified a gas limit, make sure it's not exceeded.
+        if let Some(user_limit) = handle.gas_limit() {
+            gas_limit = gas_limit.min(user_limit);
+        }
```

Comment:
I guess this comment won't be relevant if the above comment is applied, but isn't this redundant?

Gas limit just reflects what the initial limit of the call was, that never changes.
But remaining gas is set at the beginning of the execution to gas limit.